src/Microsoft.Data.SqlClient/tests/ManualTests/AlwaysEncrypted/ApiShould.cs

```diff
@@ -1983,29 +1979,38 @@ public void TestSqlCommandCancel(string connection, string value, int number)
 
                     CommandHelper.s_sleepDuringTryFetchInputParameterEncryptionInfo?.SetValue(null, true);
 
-                    Thread[] threads = new Thread[2];
+                    Task[] tasks = new Task[2];
+                    ManualResetEventSlim entryLock = new ManualResetEventSlim(false);
```

Comment:
Although this is clever, I feel like this does complicate this test significantly. I'm wondering if it makes more sense to do something simple like making the command be a WAITFOR statement that guarantees we'll have the execute task running before attempting to cancel it.

Reply:
While a WAITFOR makes perfect sense when dealing with an ExecuteScalar call, we need to test what happens if someone cancels a SqlCommand while another thread is iterating through the ExecuteReader results - we need to put that WAITFOR "in the middle of" the SELECT statement. I can't put it in a scalar-valued UDF and call that, it has a side effect.

A simple `Thread.Sleep` at the start of the cancellation thread would probably work... but that's naturally going to vary based on latency and command execution times, even on the CI legs (because these test both on local SQL Server instances and Azure SQL instances.) Too low, and thread B's cancellation logic will set the SqlCommand's internal cancellation state, just in time for thread A to call ExecuteScalar and reset that state without ever respecting the cancellation request. Too high, and thread A will have connected and completed execution before thread B cancels the command.

You're not alone in disliking the test! I used the combination of entry/exit locks and hammering on SqlCommand.Cancel as a way to make it pass most of the time, only to be met with an error in the native SNI library which looks suspiciously like memory corruption...

The operative part of this PR is purely to make sure that exceptions propagate across threads cleanly rather than terminate the process. If we're opening a can of worms, I'm perfectly happy to revert the [third commit](https://github.com/dotnet/SqlClient/pull/2968/commits/7c766b44dceee221d5c6b6eef002eb2865649fd7) and open an issue so we can look at it later.

Reply:
Hmm. Maybe I'm thinking a bit too hard about this... Looking at the original code, Thread_ExecuteReader is expecting to get a reader back, but all attempts at reading it are expected to fail. That's an even bigger race condition where the cancellation must happen between returning the metadata for the first result set and reading the first record. 

(I'm kinda struggling to fully understand the changes from the diff alone, and don't really want to load it all up in my ide right now, so forgive me if I'm restating the obvious) My thought would be to basically have this sequence

|Execute Thread|Cancel Thread|
|-----------------|---------------|
|Start thread|-|
|Execute reader|StartThread|
|-|Wait for Executed|
|Set Executed|-|
|Wait for Cancelled|-|
|-|Cancel|
|-|Set Cancelled|
|Read (and fail)|-|
|-End-|-End-|

Maybe that's exactly what you're trying to do and I'm just not seeing it. In my defense "entry" and "exit" locks I guess don't explain it well enough to me. Idk, what do you think?

Reply:
The original test isn't an ideal point of reference; I'm not actually convinced it's a valid test case at all. It essentially opens a data reader, then iterates over every row in the result set and tries to open another data reader from the same command, expecting that to throw an InvalidOperationException.

The description in my previous comment ("we need to test what happens if someone cancels a SqlCommand while another thread is iterating through the ExecuteReader results") is the behaviour I want to test, but the current code doesn't do that. I'll correct it.

Your expected cross-thread behaviour is pretty much there. The only difference is that I've swapped the order that the threads start in: the "workload" thread (whether this is ExecuteScalar or ExecuteReader) will wait until the cancellation thread signals it, rather than vice versa. I don't think it makes a real difference in this case, I just thought about the execution flow in terms of "that workload needs to stay where it is until I'm ready to start hammering the Cancel invocation".

The entry and exit locks refer to the workload thread's activity: they wait for the entry lock to be signalled before starting, then signal the exit lock to indicate that they've finished. The cancellation thread is the inverse: it signals the entry lock to indicate that it's ready for the workload thread to start, and calls SqlCommand.Cancel until the exit lock is signalled. The naming on these probably needs some work: what about calling them "StartWorkloadSignal" and "WorkloadCompleteSignal"?

Reply:
> ... I'm not actually convinced it's a valid test case at all. ...

I guess I also have no idea these tests are special to always encrypted - you'd think encryption has no bearing on cancelling a query...

> ... The only difference is that I've swapped the order that the threads start in: the "workload" thread ...

Ok, maybe we can give a little comment similar to the table I proposed just so we can have some documentation of what the expected behavior is. Otherwise I could see myself getting very lost in this code.

> ... what about calling them "StartWorkloadSignal" and "WorkloadCompleteSignal"?

Sure, I think that's a better naming scheme 👍